src/System.Linq/src/System/Linq/Select.cs

```diff
@@ -226,6 +258,17 @@ public List<TResult> ToList()
 
             public int GetCount(bool onlyIfCheap)
             {
+                // In case someone uses Count() to force evaluation of
+                // the selector, run it provided `onlyIfCheap` is false.
+
+                if (!onlyIfCheap)
+                {
+                    foreach (TSource item in _source)
+                    {
+                        _selector(item);
+                    }
+                }
```

Comment:
@JonHanna During https://github.com/dotnet/corefx/pull/12703 when I had optimized `Where.Select` and the issue of running these selectors had come up, I had originally had a `EnumerableHelpers.Count` function that `Enumerable.Count` would call after checking for Linq interfaces (just like what `ToArray` does today). This was the code I had written for the iterators

```cs
// Leave it to Count to iterate through us
public int GetCount(bool onlyIfCheap) => onlyIfCheap ? -1 : EnumerableHelpers.Count(this);
```

However, @stephentoub argued against this. See here for context: https://github.com/dotnet/corefx/pull/12703#discussion_r89671334 I ended up writing everything inline for `GetCount` in those iterators. So I just employed the same strategy here.